src/test_localization_landmark_update.cpp

```diff
@@ -0,0 +1,154 @@
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+*/
+#include <Eigen/Dense>
+#include <glog/logging.h>
+#include <gtest/gtest.h>
+
+#include "rovio/FilterStates.hpp"
+#include "rovio/LocalizationLandmarkUpdate.hpp"
+#include "rovio/MultiCamera.hpp"
+
+typedef rovio::FilterState</*nMax=*/1,/*nLevels=*/2,/*patchSize=*/20,
+    /*nCam=*/2,/*nPose=*/1,/*enableMapLocalization=*/true> FilterState;
+typedef rovio::LocalizationLandmarkUpdate<FilterState>
+LocalizationLandmarkUpdate;
+typedef typename FilterState::mtState State;
+typedef typename FilterState::mtState Noise;
+typedef typename LocalizationLandmarkUpdate::mtMeas Measurement;
+typedef typename LocalizationLandmarkUpdate::mtInnovation Innovation;
+
+TEST(LocalizationLandmarkUpdate, Jacobian) {
+  rovio::MultiCamera<State::nCam_> multi_cameras;
+  for (int cam_idx = 0u; cam_idx < State::nCam_; ++cam_idx) {
+    multi_cameras.setExtrinsics(cam_idx, V3D::Random(), QPD().setRandom());
+  }
+
+  LocalizationLandmarkUpdate update;
+  update.setCamera(&multi_cameras);
+
+  unsigned int seed = 1;
+  State state;
+  state.setRandom(seed);
+  state.updateMultiCameraExtrinsics(&multi_cameras);
+
+  // Test the update function for each camera separately.
+  for (int cam_idx = 0u; cam_idx < State::nCam_; ++cam_idx) {
+    // Generate a measurement that is in front of the camera.
+    const cv::Point2f keypoint_distorted(1.0, 2.0);
+    Eigen::Vector3d C_l;
+    CHECK(multi_cameras.cameras_[cam_idx].pixelToBearing(
+        keypoint_distorted, C_l));
+    C_l *= 10.0;
+
+    const Eigen::Vector3d C_landmark(1.0,2.0,1.0);
+    const Eigen::Vector3d M_landmark =
+        state.qCM(cam_idx).inverseRotate(C_landmark) + state.MrMC(cam_idx);
+    const Eigen::Vector3d W_landmark =
+        state.qWM().rotate(M_landmark) + state.WrWM();
+    const Eigen::Vector3d G_landmark =
+        state.qWG().inverseRotate(V3D(W_landmark - state.WrWG()));
+
+    // The keypoint measurement is disturbed to test the computation of the
+    // innovation.
+    constexpr double kKeypointDisturbance = 2.5;
+
+    Measurement measurement;
+    measurement.G_landmark() = G_landmark;
+    measurement.keypoint() << keypoint_distorted.x, keypoint_distorted.y;
+    measurement.keypoint() += Eigen::Vector2d::Constant(kKeypointDisturbance);
```

Comment:
would be nice to have a different disturbance in each direction, that might detect even more issues (e.g. swapped gradients)